src/machine.cpp

```diff
@@ -520,6 +537,54 @@ machine::machine(const std::string &dir, const machine_runtime_config &r) : mach
     }
 }
 
+void machine::poll_virtio_devices_before_select(select_fd_sets *fds, uint64_t *timeout_us) {
+    for (auto &vdev : m_vdevs) {
+        vdev->poll_before_select(fds, timeout_us);
+    }
+}
+
+bool machine::poll_virtio_devices_after_select(int select_ret, select_fd_sets *fds, i_device_state_access *da) {
+    bool interrupt_requested = false;
+    for (auto &vdev : m_vdevs) {
+        interrupt_requested |= vdev->poll_after_select(select_ret, fds, da);
+    }
+    return interrupt_requested;
+}
+
+bool machine::poll_virtio_devices(uint64_t *timeout_us, i_device_state_access *da) {
+    struct poll_virtio_upvalues {
```

Comment:
Why not simply use captures in your lambdas?

Reply:
To not trigger allocation, lambda captures will allocate memory.

Reply:
Why would a capture allocate memory? In the stack, you mean?

Reply:
I was under the impression that non trivial lambdas (with captures) could cause heap allocation when converted to an `std::function` at the moment, so I avoided them, this is not that obvious. But after some research seems like C++ is smart enough to avoid triggering heap allocation.

Reply:
I converted to lambdas with stack captures.